services/blockstorage/adapter/filesystem/file_system_persistence.go

```diff
@@ -294,44 +290,48 @@ func (f *BlockPersistence) ScanBlocks(from primitives.BlockHeight, pageSize uint
 	fromHeight := from
 	wantsMore := true
 	eof := false
-	for fromHeight <= lastBlockHeight && wantsMore && !eof {
+	for fromHeight <= inOrderHeight && wantsMore && !eof {
 		toHeight := fromHeight + primitives.BlockHeight(pageSize) - 1
-		if toHeight > lastBlockHeight {
-			toHeight = lastBlockHeight
+		if toHeight > inOrderHeight {
+			toHeight = inOrderHeight
 		}
 		page := make([]*protocol.BlockPairContainer, 0, pageSize)
+		// TODO: Gad allow update of inOrder inside page
 		for height := fromHeight; height <= toHeight; height++ {
-			initialOffset, ok := f.bhIndex.fetchBlockOffset(height)
-			if !ok {
-				return fmt.Errorf("failed to find requested block %d", uint64(from))
-			}
-			newOffset, err := file.Seek(initialOffset, io.SeekStart)
-			if newOffset != initialOffset || err != nil {
-				return errors.Wrapf(err, "failed to seek in blocks file to position %v", initialOffset)
-			}
-			aBlock, _, err := f.codec.decode(file)
-			if err != nil {
+			if aBlock, err := f.fetchBlockFromFile(height, file); err != nil {
 				if err == io.EOF || err == io.ErrUnexpectedEOF {
 					eof = true
 					break
+				} else {
+					return errors.Wrapf(err, "failed to decode block")
 				}
-				return errors.Wrapf(err, "failed to decode block")
+			} else {
+				page = append(page, aBlock)
 			}
-			page = append(page, aBlock)
 		}
 		if len(page) > 0 {
 			wantsMore = cursor(page[0].ResultsBlock.Header.BlockHeight(), page)
 		}
-		lastBlockHeight = getBlockHeight(f.bhIndex.topBlock)
-		if from <= inOrderHeight {
-			lastBlockHeight = getBlockHeight(f.bhIndex.inOrderBlock)
-		}
+		inOrderHeight = getBlockHeight(f.bhIndex.inOrderBlock)
 		fromHeight = toHeight + 1
 	}
 
 	return nil
 }
 
+func (f *BlockPersistence) fetchBlockFromFile(height primitives.BlockHeight, file *os.File) (*protocol.BlockPairContainer, error) {
```

Comment:
this method is strange - it assumes it's being called many times for the same file, but then it seeks every time which seems inefficient. if you avoid using it from `scanBlocks` you can copy it inline to it's only other call site where the file is also  opened and closed...